Sources/SourceKitLSP/Swift/DiagnosticReportManager.swift

```diff
@@ -0,0 +1,98 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2020 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+import LSPLogging
+import LanguageServerProtocol
+
+protocol DiagnosticReportManagerDelegate: AnyObject {
```

Comment:
i was considering the the way you mentioned above before, but regarding the testability of DiagnosticReportManager, i decided to have a delegate to make the testing easier.
it makes sense to remove the delegate to avoid playing ping-pong (let me update the code for it first), but i was wondering if there is a way to test it.

Reply:
I think it’s fine to test it by just sending the diagnostic requests followed by a code action request (which I think we already have a test for). That should trigger the path that computes diagnostics and also the one that re-used cached diagnostics.

Reply:
ah, okie. then, let me make this PR ready for review.